pyro/infer/mcmc/hmc.py

```diff
@@ -270,13 +270,22 @@ def _cache(self, z, potential_energy, z_grads):
         self._potential_energy_last = potential_energy
         self._z_grads_last = z_grads
 
+    def clear_cache(self):
+        self._z_last = None
+        self._potential_energy_last = None
+        self._z_grads_last = None
+
     def _fetch_from_cache(self):
         return self._z_last, self._potential_energy_last, self._z_grads_last
 
     def sample(self, params):
         z, potential_energy, z_grads = self._fetch_from_cache()
+        # recompute PE when cache is cleared
+        if z is None:
+            z = params
+            potential_energy = self.potential_fn(z)
```

Comment:
IIUC `z_grads` will still be `None` when passed into `velicity_verlet()` below. Is this intended?

Reply:
Ah I see this is ok :+1:

Reply:
Yes, the integrator will compute it if it's None.